flowmachine/flowmachine/features/subscriber/active_subscribers.py

```diff
@@ -0,0 +1,120 @@
+from flowmachine.features.subscriber.interevent_interval import IntereventInterval
+from flowmachine.features.utilities.events_tables_union import EventsTablesUnion
+
+
+class ActiveSubscribers(Query):
+    """Returns a list of subscribers active `active_days`
+    out of `interval`, optionally with at least Z call-hours active"""
+
+    # TODO: Parameterise which events tables to use + which ID method to use
+
+    def __init__(
+        self,
+        start_date: Union[date, str],
+        end_date: Union[date, str],
+        active_days: int,
+        interval: int,
+        active_hours: int = None,
+        subscriber_id: str = "msisdn",
+        events_tables: Optional[List[str]] = None,
+    ):
+        self.start_date = start_date
+        self.end_date = end_date
+        self.active_days = active_days
+        self.interval = interval
+        self.active_hours = active_hours
+        self.sub_id_column = subscriber_id
+        self._window_start = self._start_date - timedelta(days=self.interval - 1)
+        self.events_table = EventsTablesUnion(
+            start_date,
+            end_date,
+            tables=events_tables,
+            subscriber_identifier=subscriber_id,
+            columns=[subscriber_id, "datetime"],
+        )
+        super().__init__()
+
+    @property
+    def start_date(self):
+        return self._start_date.strftime("%Y-%m-%d")
+
+    @start_date.setter
+    def start_date(self, value):
+        if type(value) is str:
+            self._start_date = datetime.strptime(value, "%Y-%m-%d")
+        elif type(value) in [date, datetime]:
+            self._start_date = value
+        else:
+            raise TypeError("start_date must be datetime or yyyy-mm-dd")
+
+    @property
+    def end_date(self):
+        return self._end_date.strftime("%Y-%m-%d")
+
+    @end_date.setter
+    def end_date(self, value):
+        if type(value) is str:
+            self._end_date = datetime.strptime(value, "%Y-%m-%d")
+        elif type(value) in [date, datetime]:
+            self._end_date = value
+        else:
+            raise TypeError("end_date must be datetime or yyyy-mm-dd")
+
+    @property
+    def column_names(self) -> List[str]:
+        return ["datetime", "subscriber"]
+
+    # What should subscriber ID be?
+
+    def _make_query(self):
+
+        # Review questions:
+        # How should we pass dates around internally in fm?
+        # Should this return subscribers-day pairs, or just a list of subscribers?
+        # Should we offer the choice?
+
+        sql = f"""
+WITH ordered_events AS(
+	SELECT  subscriber, datetime
+	FROM ({self.events_table.get_query()}) AS tbl
+	WHERE datetime BETWEEN date('{self._window_start :%Y-%m-%d}') AND date('{self._end_date:%Y-%m-%d}')
+	ORDER BY subscriber, datetime
+), seen_on_days AS(
+	SELECT DISTINCT ON (event_date, subscriber)
+	subscriber, datetime::date as event_date
+	FROM ordered_events
+	ORDER BY subscriber
+), dates_of_interest AS (
```

Comment:
Checking my understanding: the 'dates_of_interest' business is so we get an output row for a subscriber who was not active on day `x`, but was active on at least `active_days` in the interval up to day `x` (which would otherwise be missed because 'seen_on_days' doesn't have a row for day `x`)?

Reply:
Correct